File: lexicon-core/ImageValue.cpp
#include "ImageValue.h"

#include <algorithm>

namespace lexicon {
namespace {
bool canonicalHash(std::string_view hash) {
  return hash.size() == 64 &&
         std::all_of(hash.begin(), hash.end(), [](char c) { return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f'); });
}
bool knownType(std::string_view mediaType) {
  return std::find(kImageMediaTypes.begin(), kImageMediaTypes.end(), mediaType) != kImageMediaTypes.end();
}
} // namespace

std::optional<ImageValue> parseImageValue(std::string_view value) {
  const auto colon = value.rfind(':');
  if (colon == std::string_view::npos)
    return std::nullopt;
  const auto mediaType = value.substr(0, colon);
  const auto hash = value.substr(colon + 1);
  if (!knownType(mediaType) || !canonicalHash(hash))
    return std::nullopt;
  return ImageValue{std::string(mediaType), std::string(hash)};
}

std::string formatImageValue(std::string_view mediaType, std::string_view hash) {
  return std::string(mediaType) + ":" + std::string(hash);
}

std::string sniffImageType(std::string_view head) {
  const auto starts = [head](std::string_view prefix, std::size_t at = 0) {
    return head.size() >= at + prefix.size() && head.substr(at, prefix.size()) == prefix;
  };
  if (starts("\x89PNG\r\n\x1a\n"))
    return "image/png";
  if (starts("\xff\xd8\xff"))
    return "image/jpeg";
  if (starts("GIF87a") || starts("GIF89a"))
    return "image/gif";
  if (starts("RIFF") && starts("WEBP", 8))
    return "image/webp";
  if (starts("BM") && head.size() >= 14)
    return "image/bmp";
  return {};
}

std::string imageExtension(std::string_view mediaType) {
  if (mediaType == "image/png") return "png";
  if (mediaType == "image/jpeg") return "jpg";
  if (mediaType == "image/gif") return "gif";
  if (mediaType == "image/webp") return "webp";
  if (mediaType == "image/bmp") return "bmp";
  return {};
}

std::string storedFileHash(FieldDataType type, std::string_view value) {
  if (type == FieldDataType::Blob)
    return std::string(value);
  if (type == FieldDataType::Image)
    if (auto image = parseImageValue(value))
      return image->hash;
  return {};
}
} // namespace lexicon

File: lexicon-web/js/imagevalue.js
// Image values: "<media type>:<SHA-256 of the file>", such as
// "image/png:6c7d...". Pure, so Node can test it; the C++ counterpart is
// lexicon-core/ImageValue.h.

// The image types Lexicon stores and shows. No SVG: it can carry script.
export const IMAGE_MEDIA_TYPES = ['image/png', 'image/jpeg', 'image/gif', 'image/webp', 'image/bmp'];

const NAMES = {
    'image/png': 'PNG', 'image/jpeg': 'JPEG', 'image/gif': 'GIF', 'image/webp': 'WebP', 'image/bmp': 'BMP',
};
const EXTENSIONS = {
    'image/png': 'png', 'image/jpeg': 'jpg', 'image/gif': 'gif', 'image/webp': 'webp', 'image/bmp': 'bmp',
};

export function parseImageValue(value) {
    const match = /^(image\/[a-z]+):([0-9a-f]{64})$/.exec(value || '');
    if (!match || !IMAGE_MEDIA_TYPES.includes(match[1])) return null;
    return { mediaType: match[1], hash: match[2] };
}

export function formatImageValue(mediaType, hash) {
    return `${mediaType}:${hash}`;
}

// The image type the bytes begin with, or '' for anything else. The first 16
// bytes are enough.
export function sniffImageType(bytes) {
    const starts = (prefix, at = 0) => bytes.length >= at + prefix.length
        && prefix.every((byte, index) => bytes[at + index] === byte);
    const ascii = (text) => [...text].map((ch) => ch.charCodeAt(0));
    if (starts([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a])) return 'image/png';
    if (starts([0xff, 0xd8, 0xff])) return 'image/jpeg';
    if (starts(ascii('GIF87a')) || starts(ascii('GIF89a'))) return 'image/gif';
    if (starts(ascii('RIFF')) && starts(ascii('WEBP'), 8)) return 'image/webp';
    if (starts(ascii('BM')) && bytes.length >= 14) return 'image/bmp';
    return '';
}

// "PNG image", or '' when the value is no image value.
export function describeImage(value) {
    const image = parseImageValue(value);
    return image ? `${NAMES[image.mediaType]} image` : '';
}

// "<name>.<extension>" for Download, without characters file systems refuse.
export function imageFileName(name, value) {
    const image = parseImageValue(value);
    const base = (name || '').trim().replace(/[\\/:*?"<>|]/g, '_') || 'image';
    return image ? `${base}.${EXTENSIONS[image.mediaType]}` : base;
}
